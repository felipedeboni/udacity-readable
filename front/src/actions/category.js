export const FETCH_CATEGORIES = 'FETCH_CATEGORIES';
export const FETCH_CATEGORIES_SUCCESS = 'FETCH_CATEGORIES_SUCCESS';
export const FETCH_CATEGORIES_FAIL = 'FETCH_CATEGORIES_FAIL';

export const fetchCategories = () => {
  return {
    type: FETCH_CATEGORIES,
    payload: {
      request: {
        url: '/categories'
      }
    }
  };
};
