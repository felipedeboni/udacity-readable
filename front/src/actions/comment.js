import uuid from 'utils/uuid';

// GET /posts/:id/comments
// Get all the comments for a single post.
export const FETCH_COMMENTS_BY_POST = 'FETCH_COMMENTS_BY_POST';
export const FETCH_COMMENTS_BY_POST_SUCCESS = 'FETCH_COMMENTS_BY_POST_SUCCESS';
export const FETCH_COMMENTS_BY_POST_FAIL = 'FETCH_COMMENTS_BY_POST_FAIL';
export const fetchCommentsByPost = postId => {
  return {
    type: FETCH_COMMENTS_BY_POST,
    payload: {
      request: {
        url: `/posts/${postId}/comments`
      }
    }
  };
};

// GET /comments/:id
// Get the details for a single comment.
export const FETCH_COMMENT = 'FETCH_COMMENT';
export const FETCH_COMMENT_SUCCESS = 'FETCH_COMMENT_SUCCESS';
export const FETCH_COMMENT_FAIL = 'FETCH_COMMENT_FAIL';
export const fetchComment = id => {
  return {
    type: FETCH_COMMENT,
    payload: {
      request: {
        url: `/comments/${id}`
      }
    }
  };
};

// POST /comments
// Add a comment to a post.
export const ADD_COMMENT = 'ADD_COMMENT';
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS';
export const ADD_COMMENT_FAIL = 'ADD_COMMENT_FAIL';
export const addComment = (parentId, values) => {
  const { author, body } = values;
  const data = {
    id: uuid(),
    timestamp: Date.now(),
    parentId,
    author,
    body
  };

  return {
    type: ADD_COMMENT,
    payload: {
      request: {
        method: 'POST',
        url: '/comments',
        data
      }
    }
  };
};

// PUT /comments/:id
// Edit the details of an existing comment.
export const EDIT_COMMENT = 'EDIT_COMMENT';
export const EDIT_COMMENT_SUCCESS = 'EDIT_COMMENT_SUCCESS';
export const EDIT_COMMENT_FAIL = 'EDIT_COMMENT_FAIL';
export const editComment = (id, data) => {
  return {
    type: EDIT_COMMENT,
    payload: {
      request: {
        method: 'PUT',
        url: `/comments/${id}`,
        data
      }
    }
  };
};

// DELETE /comments/:id
// Sets a comment's deleted flag to true.
export const REMOVE_COMMENT = 'REMOVE_COMMENT';
export const REMOVE_COMMENT_SUCCESS = 'REMOVE_COMMENT_SUCCESS';
export const REMOVE_COMMENT_FAIL = 'REMOVE_COMMENT_FAIL';
export const removeComment = id => {
  return {
    type: REMOVE_COMMENT,
    payload: {
      request: {
        method: 'DELETE',
        url: `/comments/${id}`
      }
    }
  };
};

// POST /comments/:id
// Used for voting on a comment.
export const VOTE_COMMENT = 'VOTE_COMMENT';
export const VOTE_COMMENT_SUCCESS = 'VOTE_COMMENT_SUCCESS';
export const VOTE_COMMENT_FAIL = 'VOTE_COMMENT_FAIL';
const vote = (id, type) => {
  return {
    type: VOTE_COMMENT,
    payload: {
      request: {
        method: 'POST',
        url: `/comments/${id}`,
        data: {
          option: type
        }
      }
    }
  };
};

export const voteCommentUp = id => {
  return vote(id, 'upVote');
};

export const voteCommentDown = id => {
  return vote(id, 'downVote');
};
