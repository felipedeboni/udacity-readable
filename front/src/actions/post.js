import uuid from 'utils/uuid';

// GET /posts
// Get all of the posts. Useful for the main page when no category is selected.
export const FETCH_ALL_POSTS = 'FETCH_ALL_POSTS';
export const FETCH_ALL_POSTS_SUCCESS = 'FETCH_ALL_POSTS_SUCCESS';
export const FETCH_ALL_POSTS_FAIL = 'FETCH_ALL_POSTS_FAIL';
export const fetchAllPosts = () => {
  return {
    type: FETCH_ALL_POSTS,
    payload: {
      request: {
        url: `/posts`
      }
    }
  };
};

// GET /:category/posts
// Get all of the posts for a particular category.
export const FETCH_POSTS_BY_CATEGORY = 'FETCH_POSTS_BY_CATEGORY';
export const FETCH_POSTS_BY_CATEGORY_SUCCESS =
  'FETCH_POSTS_BY_CATEGORY_SUCCESS';
export const FETCH_POSTS_BY_CATEGORY_FAIL = 'FETCH_POSTS_BY_CATEGORY_FAIL';
export const fetchPostsByCategory = categoryId => {
  return {
    type: FETCH_POSTS_BY_CATEGORY,
    payload: {
      request: {
        url: `/${categoryId}/posts`
      }
    }
  };
};

// GET /posts/:id
// Get the details of a single post.
export const FETCH_POST = 'FETCH_POST';
export const FETCH_POST_SUCCESS = 'FETCH_POST_SUCCESS';
export const FETCH_POST_FAIL = 'FETCH_POST_FAIL';
export const fetchPost = id => {
  return {
    type: FETCH_POST,
    payload: {
      request: {
        url: `/posts/${id}`
      }
    }
  };
};

// POST /posts
// Add a new post.
export const ADD_POST = 'ADD_POST';
export const ADD_POST_SUCCESS = 'ADD_POST_SUCCESS';
export const ADD_POST_FAIL = 'ADD_POST_FAIL';
export const addPost = ({ category, author, title, body }) => {
  const data = {
    id: uuid(),
    timestamp: Date.now(),
    category,
    author,
    title,
    body
  };

  return {
    type: ADD_POST,
    payload: {
      request: {
        method: 'POST',
        url: '/posts',
        data
      }
    }
  };
};

// PUT /posts/:id
// Edit the details of an existing post.
export const EDIT_POST = 'EDIT_POST';
export const EDIT_POST_SUCCESS = 'EDIT_POST_SUCCESS';
export const EDIT_POST_FAIL = 'EDIT_POST_FAIL';
export const editPost = (id, data) => {
  return {
    type: EDIT_POST,
    payload: {
      request: {
        method: 'PUT',
        url: `/posts/${id}`,
        data
      }
    }
  };
};

// DELETE /posts/:id
// Sets the deleted flag for a post to 'true'.
// Sets the parentDeleted flag for all child comments to 'true'.
export const REMOVE_POST = 'REMOVE_POST';
export const REMOVE_POST_SUCCESS = 'REMOVE_POST_SUCCESS';
export const REMOVE_POST_FAIL = 'REMOVE_POST_FAIL';
export const removePost = id => {
  return {
    type: REMOVE_POST,
    payload: {
      id: id,
      request: {
        method: 'DELETE',
        url: `/posts/${id}`
      }
    }
  };
};

// POST /posts/:id
// Used for voting on a post.
export const VOTE_POST = 'VOTE_POST';
export const VOTE_POST_SUCCESS = 'VOTE_POST_SUCCESS';
export const VOTE_POST_FAIL = 'VOTE_POST_FAIL';
const vote = (id, type) => {
  return {
    type: VOTE_POST,
    payload: {
      request: {
        method: 'POST',
        url: `/posts/${id}`,
        data: {
          option: type
        }
      }
    }
  };
};

export const votePostUp = id => {
  return vote(id, 'upVote');
};

export const votePostDown = id => {
  return vote(id, 'downVote');
};
