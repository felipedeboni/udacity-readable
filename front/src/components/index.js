export { default as App } from './App/App';

export { default as Header } from './Header/Header';

export { default as CommentForm } from './CommentForm/CommentForm';
export { default as CommentsList } from './CommentsList/CommentsList';
export { default as CommentListItem } from './CommentListItem/CommentListItem';

export { default as Post } from './Post/Post';
export { default as PostForm } from './PostForm/PostForm';
export { default as PostsList } from './PostsList/PostsList';
export { default as PostListItem } from './PostListItem/PostListItem';

export { default as Score } from './Score/Score';

export { default as PageNotFound } from './PageNotFound/PageNotFound';
