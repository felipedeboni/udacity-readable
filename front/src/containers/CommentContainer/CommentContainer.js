import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  fetchComment,
  removeComment,
  FETCH_COMMENT_FAIL
} from 'actions/comment';

class CommentContainer extends Component {
  componentDidMount() {
    const { commentId, comment, fetch } = this.props;
    if (!comment) {
      fetch(commentId)
        .then(this.afterFetch)
        .catch(this.afterFetch);
    }
  }

  afterFetch = response => {
    const { history } = this.props;
    if (
      response.type === FETCH_COMMENT_FAIL ||
      !response.payload ||
      !response.payload.data.id
    ) {
      history.replace('/404');
    }
  };

  render() {
    const { component, comment, remove } = this.props;

    if (comment) {
      return React.createElement(component, {
        comment: comment,
        removeHandler: remove,
        ...this.props
      });
    }

    return null;
  }
}

const mapProps = (state, ownProps) => {
  return {
    comment: ownProps.comment || state.comments[ownProps.commentId]
  };
};

const mapDispatch = dispatch => {
  return {
    fetch: commentId => dispatch(fetchComment(commentId)),
    remove: commentId => dispatch(removeComment(commentId))
  };
};

export default connect(mapProps, mapDispatch)(CommentContainer);
