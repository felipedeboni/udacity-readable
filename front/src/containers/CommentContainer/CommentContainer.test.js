import React from 'react';
import * as CommentActions from 'actions/comment';
import CommentContainer from './CommentContainer';
import { CommentListItem } from 'components';

describe('CommentContainer', () => {
  const render = commentId => {
    const [comment] = global.mock.comments;
    const state = {
      comments: global.mock.commentsObject
    };

    const { component, store } = global.shallowWithStore(
      <CommentContainer
        commentId={commentId || comment.id}
        component={CommentListItem}
      />,
      state
    );

    return {
      component,
      store,
      comment
    };
  };

  it('renders without crashing', () => {
    const { component } = render();

    expect(component.dive()).toMatchSnapshot();
  });

  it('fetches the comment', () => {
    const [, comment] = global.mock.comments;

    const { component, store } = global.shallowWithStore(
      <CommentContainer
        commentId={comment.id}
        component={CommentListItem}
        history={{ push: jest.fn(), replace: jest.fn() }}
      />,
      { comments: {} }
    );

    expect(component).toMatchSnapshot();
    component.dive();
    expect(
      store.getActions().find(_ => _.type === CommentActions.FETCH_COMMENT)
    ).toBeDefined();
  });

  it('dispatch removeComment', () => {
    const { component, store } = render();
    const form = component;

    form.props().remove({});
    expect(
      store.getActions().find(_ => _.type === CommentActions.REMOVE_COMMENT)
    ).toBeDefined();
  });
});
