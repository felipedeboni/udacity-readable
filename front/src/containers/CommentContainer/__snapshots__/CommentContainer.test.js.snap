// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentContainer fetches the comment 1`] = `
<CommentContainer
  commentId="8tu4bsun805n8un48ve89"
  component={[Function]}
  fetch={[Function]}
  history={
    Object {
      "push": [Function],
      "replace": [Function],
    }
  }
  remove={[Function]}
/>
`;

exports[`CommentContainer renders without crashing 1`] = `
<CommentListItem
  comment={
    Object {
      "author": "thingtwo",
      "body": "Hi there! I am a COMMENT.",
      "deleted": false,
      "id": "894tuq4ut84ut8v4t8wun89g",
      "parentDeleted": false,
      "parentId": "8xf0y6ziyjabvozdd253nd",
      "timestamp": 1468166872634,
      "voteScore": 6,
    }
  }
  commentId="894tuq4ut84ut8v4t8wun89g"
  component={[Function]}
  fetch={[Function]}
  remove={[Function]}
  removeHandler={[Function]}
/>
`;
